pints/_mcmc/_nuts.py

```diff
@@ -0,0 +1,571 @@
+    """
+
+    # intialise at epsilon = 1.0 (shouldn't matter where we start)
+    epsilon = 1.0
+    r = np.random.normal(size=len(theta))
+    hamiltonian = L - 0.5 * r.dot(r)
+    L_dash, grad_L_dash, theta_dash, r_dash = \
+        yield from leapfrog(theta, L, grad_L, r, epsilon, step_size)
+    hamiltonian_dash = L_dash - 0.5 * r_dash.dot(r_dash)
+    comparison = hamiltonian_dash - hamiltonian
+
+    # determine whether we are doubling or halving
+    alpha = 2 * int(comparison > np.log(0.5)) - 1
+
+    # double or half epsilon until acceptance probability crosses 0.5
+    while comparison * alpha > np.log(2) * (-alpha):
+        epsilon = 2**alpha * epsilon
+        L_dash, grad_L_dash, theta_dash, r_dash = \
+            yield from leapfrog(theta, L, grad_L, r, epsilon, step_size)
+        hamiltonian_dash = L_dash - 0.5 * r_dash.dot(r_dash)
+        comparison = hamiltonian_dash - hamiltonian
+    return epsilon
+
+
+@asyncio.coroutine
+def nuts_sampler(x0, delta, M_adapt, step_size):
+    """
+    The dual averaging NUTS mcmc sampler given in Algorithm 6 of [1].
+
+    Implemented as a coroutine that continually generates new theta values to
+    evaluate (L, L') at. Users must send (L, L') back to the coroutine to
+    continue execution. The end of an mcmc step is signalled by generating a
+    tuple of values (theta, L, acceptance probability, number of leapfrog
+    steps)
+
+
+    References
+    ----------
+    .. [1] Hoffman, M. D., & Gelman, A. (2014). The No-U-Turn sampler:
+           adaptively setting path lengths in Hamiltonian Monte Carlo.
+           Journal of Machine Learning Research, 15(1), 1593-1623.
+
+    """
+    # Initialise sampler with x0 and calculate logpdf
+    theta = x0
+    L, grad_L = (yield theta)
+
+    # Check first point is somewhere sensible
+    if not np.isfinite(L):
+        raise ValueError(
+            'Initial point for MCMC must have finite logpdf.')
+
+    # find a good value to start epsilon at
+    # (this will later be refined so that the acceptance probability matches
+    # delta)
+    epsilon = yield from find_reasonable_epsilon(theta, L, grad_L, step_size)
+
+    # default values taken from [1]
+    mu = np.log(10 * epsilon)
+    log_epsilon_bar = np.log(1)
+    H_bar = 0
+    gamma = 0.05
+    t0 = 10
+    kappa = 0.75
+
+    # start at iteration 1
+    m = 1
+
+    # provide an infinite generator of mcmc steps....
+    while True:
+        # randomly sample momentum
+        r0 = np.random.normal(size=len(theta))
+        hamiltonian0 = L - 0.5 * r0.dot(r0)
+
+        # use slice sampling
+        log_u = np.log(np.random.uniform(0, 1)) + hamiltonian0
+
+        # create initial integration path state
+        state = NutsState(theta, r0, L, grad_L, 1, 1, None, None)
+        j = 0
+
+        # build up an integration path with 2^j points, stopping when we either
+        # encounter a U-Turn, or reach a max number of points 2^10
+        while j < 10 and state.s == 1:
```

Comment:
@martinjrobins Whilst I remember, can we make the tree depth a hyper-parameter?

Reply:
done